lib/spack/spack/cmd/buildcache.py

```diff
@@ -790,18 +820,157 @@ def update_index(mirror: spack.mirrors.mirror.Mirror, update_keys=False):
         spack.binary_distribution._url_generate_package_index(url, tmpdir)
 
     if update_keys:
-        try:
-            with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpdir:
-                spack.binary_distribution.generate_key_index(url, tmpdir)
-        except spack.binary_distribution.CannotListKeys as e:
-            # Do not error out if listing keys went wrong. This usually means that the _gpg path
-            # does not exist. TODO: distinguish between this and other errors.
-            tty.warn(f"did not update the key index: {e}")
+        mirror_update_keys(mirror)
+
+
+def mirror_update_keys(mirror: spack.mirrors.mirror.Mirror):
+    url = mirror.push_url
+    try:
+        with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpdir:
+            spack.binary_distribution.generate_key_index(url, tmpdir)
+    except spack.binary_distribution.CannotListKeys as e:
+        # Do not error out if listing keys went wrong. This usually means that the _gpg path
+        # does not exist. TODO: distinguish between this and other errors.
+        tty.warn(f"did not update the key index: {e}")
+
+
+def read_concrete_hashes(source: str) -> List[str]:
+    """Read all of the concrete hashes from a given source"""
+
+    # Try to read input as a spec
+    try:
+        concrete_specs = spack.cmd.parse_specs(source, concretize=True)
+        return [spec.dag_hash() for spec in concrete_specs]
+    except Exception:
+        pass
+
+    # Try to read input as a lockfile
+    try:
+        if not os.path.exists(source):
+            raise FileNotFoundError(f"Expected lockfile: {source}")
+
+        with tempfile.TemporaryDirectory(dir=spack.stage.get_stage_root()) as tmpdir:
+            # Touch a dummy manifest file
+            with open(os.path.join(tmpdir, "spack.yaml"), "w", encoding="utf-8") as fd:
+                fd.write("spack: {}")
+            env = spack.environment.Environment(tmpdir)
+            with open(source, "r", encoding="utf-8") as fd:
+                env._read_lockfile(fd)
+            return list(env.specs_by_hash.keys())
+    except Exception as e:
+        raise spack.error.SpackError(f"Could not determine spec source type of {source}") from e
+
+
+def update_view(
+    mirror: spack.mirrors.mirror.Mirror,
+    update_mode: ViewUpdateMode,
+    *sources: str,
+    name: Optional[str] = None,
+    update_keys: bool = False,
+):
+    """update a buildcache view index"""
+    # OCI images do not support views.
+    print(f"mirror: {mirror.name}")
+    print(f"mode: {update_mode}")
+    print(f"name: {name}")
+    for source in sources:
+        print(f"source: {source}")
```

Comment:
This looks like vestigial debugging code. If you want to keep it, we should turn it into a `tty.msg` or `tty.debug` call.